homeassistant/components/influxdb/__init__.py

```diff
@@ -478,42 +481,95 @@ def close_v1():
+    """Set up InfluxDB from a config entry."""
+    data = entry.data
 
+    hass_config = await conf_util.async_hass_config_yaml(hass)
+
+    influx_yaml = CONFIG_SCHEMA(hass_config).get(DOMAIN, {})
+    default_yaml: dict[str, Any] = {
+        "entity_globs": [],
+        "entities": [],
+        "domains": [],
+    }
+
+    options = {
+        CONF_RETRY_COUNT: influx_yaml.get(CONF_RETRY_COUNT, 0),
+        CONF_PRECISION: influx_yaml.get(CONF_PRECISION),
+        CONF_MEASUREMENT_ATTR: influx_yaml.get(
+            CONF_MEASUREMENT_ATTR, DEFAULT_MEASUREMENT_ATTR
+        ),
+        CONF_DEFAULT_MEASUREMENT: influx_yaml.get(CONF_DEFAULT_MEASUREMENT),
+        CONF_OVERRIDE_MEASUREMENT: influx_yaml.get(CONF_OVERRIDE_MEASUREMENT),
+        CONF_INCLUDE: influx_yaml.get(CONF_INCLUDE, default_yaml),
+        CONF_EXCLUDE: influx_yaml.get(CONF_EXCLUDE, default_yaml),
+        CONF_TAGS: influx_yaml.get(CONF_TAGS, {}),
+        CONF_TAGS_ATTRIBUTES: influx_yaml.get(CONF_TAGS_ATTRIBUTES, []),
+        CONF_IGNORE_ATTRIBUTES: influx_yaml.get(CONF_IGNORE_ATTRIBUTES, []),
+        CONF_COMPONENT_CONFIG: influx_yaml.get(CONF_COMPONENT_CONFIG, {}),
+        CONF_COMPONENT_CONFIG_DOMAIN: influx_yaml.get(CONF_COMPONENT_CONFIG_DOMAIN, {}),
+        CONF_COMPONENT_CONFIG_GLOB: influx_yaml.get(CONF_COMPONENT_CONFIG_GLOB, {}),
+    }
+
+    config = data | options
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the InfluxDB component."""
-    conf = config[DOMAIN]
     try:
-        influx = get_influx_connection(conf, test_write=True)
-    except ConnectionError as exc:
-        _LOGGER.error(RETRY_MESSAGE, exc)
-        event_helper.call_later(
-            hass, RETRY_INTERVAL, lambda _: _retry_setup(hass, config)
-        )
-        return True
+        influx = await hass.async_add_executor_job(get_influx_connection, config, True)
+    except ConnectionError as err:
+        raise ConfigEntryNotReady(err) from err
+
+    event_to_json = _generate_event_to_json(config)
+    max_tries = config.get(CONF_RETRY_COUNT)
+    influx_thread = hass.data[DOMAIN] = InfluxThread(
+        hass, entry, influx, event_to_json, max_tries
+    )
+    influx_thread.start()
 
-    event_to_json = _generate_event_to_json(conf)
-    max_tries = conf.get(CONF_RETRY_COUNT)
-    instance = hass.data[DOMAIN] = InfluxThread(hass, influx, event_to_json, max_tries)
-    instance.start()
+    entry.runtime_data = influx_thread
 
-    def shutdown(event):
-        """Shut down the thread."""
-        instance.queue.put(None)
-        instance.join()
-        influx.close()
+    return True
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+
+async def async_unload_entry(hass: HomeAssistant, entry: InfluxDBConfigEntry) -> bool:
+    """Unload a config entry."""
+    influx_thread = entry.runtime_data
+
+    influx_thread.shutdown()
+
+    return True
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
```

Comment:
Let's have async_setup in the top

Reply:
The top of the file? You mean above `async_setup_entry`? Or all the way?

Reply:
above async_setup_entry